pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go

```
@@ -0,0 +1,477 @@
+	derivedEpSlice.Ports = shallowCopyServicePortToEndpointPort(derivedService.Spec.Ports)
+}
+
+func shallowCopyServicePortToEndpointPort(ports []corev1.ServicePort) []discoveryv1.EndpointPort {
+	endpointPorts := make([]discoveryv1.EndpointPort, 0, len(ports))
+	for _, port := range ports {
+		endpointPort := discoveryv1.EndpointPort{
+			Port:        &port.Port,
+			AppProtocol: port.AppProtocol,
+		}
+		if port.Name != "" {
+			endpointPort.Name = &port.Name
+		}
+		if port.Protocol != "" {
+			endpointPort.Protocol = &port.Protocol
+		}
+		endpointPorts = append(endpointPorts, endpointPort)
+	}
+	return endpointPorts
+}
+
+func (r *mcsAPIEndpointSliceMirrorReconciler) newDerivedEndpointSlice(
+	localEpSlice *discoveryv1.EndpointSlice, derivedService *corev1.Service,
+) *discoveryv1.EndpointSlice {
+	derivedEndpointSlice := discoveryv1.EndpointSlice{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      derivedService.Name + "-" + getSuffix(localEpSlice),
+			Namespace: derivedService.Namespace,
+		},
+	}
+
+	r.updateDerivedEndpointSlice(&derivedEndpointSlice, localEpSlice, derivedService)
+	return &derivedEndpointSlice
+}
+
+func (r *mcsAPIEndpointSliceMirrorReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var epSlice discoveryv1.EndpointSlice
+	if err := r.Client.Get(ctx, req.NamespacedName, &epSlice); err != nil {
+		return controllerruntime.Fail(client.IgnoreNotFound(err))
+	}
+
+	var localEpSlice *discoveryv1.EndpointSlice
+	var derivedEpSlice *discoveryv1.EndpointSlice
+	var derivedService *corev1.Service
+	var err error
+
+	if epSlice.Labels[discoveryv1.LabelManagedBy] == endpointSliceLocalMCSAPIControllerName {
+		derivedEpSlice = &epSlice
+		derivedService, err = r.getDerivedService(
+			ctx,
+			types.NamespacedName{
+				Name:      derivedEpSlice.Labels[discoveryv1.LabelServiceName],
+				Namespace: derivedEpSlice.Namespace,
+			},
+		)
+		if err != nil || derivedService == nil {
```

Nit. Rather than checking `derivedService == nil`, I would propagate the NotFound error here as well (directly calling`client.Get` to get rid of the helper) , and then ignore it below. IMO that makes the intent a bit clearer.